horovod/common/ops/hip_operations.cc

```diff
@@ -34,18 +35,39 @@ class GPUContext::impl {
     auto& mutex = hip_events_mutex;
     {
       std::lock_guard<std::mutex> guard(mutex);
-      auto& queue = hip_events[device];
+      auto key = std::make_pair(device, stream);
+      auto& queue = hip_events[key];
+      if (!prepopulated[key]) {
+        // On first call for device and stream pair, prepopulate event queue.
+        // This is to minimize event reuse of callback events passed to
+        // framework.
+        for (int i = 0; i < N_HIP_EVENTS_PREPOPULATE; ++i) {
+          hipEvent_t ev;
+          status = hipEventCreateWithFlags(&ev, hipEventDisableTiming);
+          queue.emplace(std::make_shared<hipEvent_t>(ev), stream);
+        }
+        prepopulated[key] = true;
+      }
       if (!queue.empty()) {
         *event = queue.front();
+        event->event_idx = ++hip_event_idx[key];
         queue.pop();
         return hipSuccess;
       }
     }
 
-    return hipEventCreateWithFlags(event, hipEventDisableTiming);
+    hipEvent_t ev;
+    status = hipEventCreateWithFlags(&ev, hipEventDisableTiming);
+    event->event = std::make_shared<hipEvent_t>(ev);
+    event->stream = stream;
+    auto key2 = std::make_pair(device, stream);
+    event->event_idx = ++hip_event_idx[key2];
+
+
```

Comment:
I think here is no need for second empty line.

Reply:
> @weihanmines any updates on how we could test continuously on a ROCM system?

Hi Enrico @EnricoMi, I am working with my colleagues to set up external CI for ROCM on our machine. I will keep you posted on this.